src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -0,0 +1,444 @@
+                FlowPiece existingPiece = branches.get(nextBlockStart);
+                if (existingPiece == null && myPiece instanceof FlowSegment) { // No merge, just add to segment
+                    ((FlowSegment) myPiece).add(nextBlockStart);
+                    branches.put(nextBlockStart, myPiece);
+                } else if (existingPiece == null && myPiece instanceof Fork) {  // No merge, we had a fork. Start a segment preceding the fork
+                    FlowSegment newSegment = new FlowSegment();
+                    newSegment.isLeaf = false;
+                    newSegment.add(nextBlockStart);
+                    newSegment.after = myPiece;
+                    pieceIterator.remove();
+                    pieceIterator.add(newSegment);
+                    branches.put(nextBlockStart, newSegment);
+                } else if (existingPiece != null) {  // Always not null. We're merging into another thing, we're going to elliminate a branch
+                    if (existingPiece instanceof Fork) {
+                        ((Fork) existingPiece).following.add(myPiece);
+                    } else { // Split a flow segment so it forks against this one
+                        Fork f = ((FlowSegment) existingPiece).split(branches, (BlockStartNode)nextBlockStart, myPiece);
+                        // If we split the existing segment at its end, we created a fork replacing its latest node
+                        // Thus we must replace the piece with the fork ahead of it
+                        if (f.following.contains(existingPiece) ) {
+                            int headIndex = livePieces.indexOf(existingPiece);
+                            livePieces.set(headIndex, f);
+                        }
+                        parallelForks.add(f);
+                    }
+
+                    // Merging removes the piece & its iterator from heads
+                    itIterator.remove();
+                    pieceIterator.remove();
+                }
+            }
+        }
+
+        // If we hit issues with the ordering of blocks by depth, apply a sorting to the parallels by depth
+        return convertForksToBlockStarts(parallelForks);
+    }
+
+    @Override
+    protected void setHeads(@Nonnull Collection<FlowNode> heads) {
+        if (heads.size() > 1) {
+            //throw new IllegalArgumentException("ForkedFlowScanner can't handle multiple head nodes yet");
+            parallelBlockStartStack = leastCommonAncestor(new LinkedHashSet<FlowNode>(heads));
+            currentParallelStart = parallelBlockStartStack.pop();
+            currentParallelStartNode = currentParallelStart.forkStart;
+            myCurrent = currentParallelStart.unvisited.pop();
+            myNext = myCurrent;
+            currentParallelStart.remainingBranches--;
+            walkingFromFinish = false;
+        } else {
+            FlowNode f = heads.iterator().next();
+            walkingFromFinish = f instanceof FlowEndNode;
```

Comment:
Is it realistic that we have a single FlowEndNode head? Looks like a broken graph